### PR TITLE
Fix diagnostics in multi file projects

### PR DESCRIPTION
**Pull request by sdankel**
## Description

Closes https://github.com/FuelLabs/sway/issues/5027

Now instead of caching the `Diagnostics` object, we cache a mapping of file path -> `Diagnostics`

When we publish diagnostics, we only publishing the diagnostics for the given URI, not for the entire project. I added a test where there are no diagnostics in the main.sw file, but a bunch of diagnostics in the other files in the project.

## Checklist

- [x] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-lsp/src/core/session.rs [hunk 1]
```diff
@@ -445,12 +446,10 @@ pub fn parse_project(uri: &Url) -> Result<ParseResult, LanguageServerError> {
...
     for (i, (value, handler)) in results.into_iter().enumerate() {
         // We can convert these destructured elements to a Vec<Diagnostic> later on.
-        let (errors, warnings) = handler.consume();
+        let current_diagnostics = handler.consume();
+        diagnostics = current_diagnostics;
 
         if value.is_none() {
-            // If there was an unrecoverable error in the parser
-            // make sure to still return the diagnostics.
-            diagnostics = get_diagnostics(&warnings, &errors);
```
Comment (JoshuaBatty) [inline sway-lsp/src/core/session.rs:453]: do we not want to still capture diagnostics here?

**Comment by JoshuaBatty** [comment]
Just tested this locally, wow I can't believe how bad it was! We need to get this in the beta-4 release for sure! Thanks for spotting this error! So much nicer now.

File: sway-lsp/src/core/session.rs (hunk 1 above)
Comment (JoshuaBatty) [reply sway-lsp/src/core/session.rs:453]: ok cool yep that makes sense. thanks.

**Review by JoshuaBatty — APPROVED** [review]
Awesome work

**Review by cr-fuel — APPROVED** [review]

**Pull request merged**